velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -224,60 +297,91 @@ HiveDataSink::HiveDataSink(
+          HiveConfig::maxPartitionsPerWriters(connectorQueryCtx_->config())),
       partitionChannels_(getPartitionChannels(insertTableHandle_)),
       partitionIdGenerator_(
           !partitionChannels_.empty() ? std::make_unique<PartitionIdGenerator>(
                                             inputType_,
                                             partitionChannels_,
-                                            HiveConfig::maxPartitionsPerWriters(
-                                                connectorQueryCtx_->config()),
+                                            maxOpenWriters_,
                                             connectorQueryCtx_->memoryPool())
-                                      : nullptr) {
+                                      : nullptr),
+      bucketCount_(
+          insertTableHandle_->bucketProperty() == nullptr
+              ? 0
+              : insertTableHandle_->bucketProperty()->bucketCount()),
+      bucketFunction_(
+          isBucketed() ? createBucketFunction(
+                             insertTableHandle_->bucketProperty(),
+                             inputType_)
+                       : nullptr) {
+  VELOX_USER_CHECK(
+      !isBucketed() || isPartitioned(),
+      "A bucket table must be partitioned: {}");
+  if (isBucketed()) {
+    VELOX_USER_CHECK_LT(
+        bucketCount_,
+        maxBucketCount(),
+        "bucketCount {} must be smaller than {}",
+        bucketCount_,
+        maxBucketCount());
+  }
+  VELOX_USER_CHECK(
+      (commitStrategy_ == CommitStrategy::kNoCommit) ||
+          (commitStrategy_ == CommitStrategy::kTaskCommit),
+      "Unsupported commit strategy: {}",
+      commitStrategyToString(commitStrategy_));
   // TODO: remove this hack after Prestissimo adds to register dwrf writer.
   facebook::velox::dwrf::registerDwrfWriterFactory();
 }
 
 void HiveDataSink::appendData(RowVectorPtr input) {
   // Write to unpartitioned table.
-  if (partitionChannels_.empty()) {
-    ensureSingleWriter();
-
-    writers_[0]->write(input);
-    writerInfo_[0]->numWrittenRows += input->size();
+  if (!isPartitioned()) {
+    const auto index = ensureWriter(HiveWriterId::unpartitionedId());
+    writers_[index]->write(input);
+    writerInfo_[index]->numWrittenRows += input->size();
     return;
   }
 
   // Write to partitioned table.
-  partitionIdGenerator_->run(input, partitionIds_);
-
-  ensurePartitionWriters();
+  computePartitionAndBucketIds(input);
 
+  // Lazy load all the input columns.
   for (column_index_t i = 0; i < input->childrenSize(); ++i) {
     input->childAt(i)->loadedVector();
   }
 
-  const auto numPartitions = partitionIdGenerator_->numPartitions();
-
   // All inputs belong to a single partition.
-  if (numPartitions == 1) {
-    writers_[0]->write(input);
-    writerInfo_[0]->numWrittenRows += input->size();
+  if (!isBucketed() && partitionIdGenerator_->numPartitions() == 1) {
+    const auto index = ensureWriter(HiveWriterId{0});
```

Comment:
Why are we passing 0 for partition ID?

Reply:
Because it is a special case for non-bucketed partition table and all the inputs belong to a single partition and we skip the split step below and the partition id must be zero.